### DISCUSSION TITLE
Out of memory error when reading many csv files using read_csv_auto

### DISCUSSION THREAD

**Discussion by albahhar-ahmed** [duckdb/duckdb#7463] [Q&A]
hello,

I have a lot of csv files (> 1 million) that follow /year/month/day folder structure and they don't fit in memory (> 100 GB); they have the same schema as well. I'm trying to read all of them and create a table but its seems I'm running out of memory. I'm using the following command

```sql
CREATE TABLE tbl AS SELECT * FROM read_csv_auto('root/**/**/**/*.csv');
```
which results in `Error: Out of Memory Error: Failed to allocate block of 32768 bytes`

is there a way to resolve this? thanks :pray:

**Comment by Mytherin** [comment]
Are you trying to load them into an in-memory database, or into a persistent database?

**Reply by albahhar-ahmed** [reply]
I tried both. but the goal is to store in a persistent database.

**Reply by albahhar-ahmed** [reply]
is this expected and is a missing features or is it a bug ? Are there any workarounds for now to fix this ?

**Reply by Mytherin** [reply]
It is not expected for the system to run out of memory when streaming CSV files into a persistent database. Could you verify the issue is still there on the latest master as well? We have made some changes to data loading/CSV parsing recently.

**Reply by albahhar-ahmed** [reply]
I tried v0.7.2-dev3666 eae707d54c
I get `Error: IO Error: Cannot use multiple '**' in one path`

**Reply by lverdoes** [reply]
Have you tried to use the following path: `root/**/*.csv` ? It should find the files in the subdirectories

**Reply by albahhar-ahmed** [reply]
I tried it and it worked for a few files. However, for all files, it takes so long without any progress (> 1 hour) so I killed the process without any success. note: I moved the files to external storage so a slow down is expected but this seems way too slow.

**Reply by albahhar-ahmed** [reply]
I tried again on SSD for only 1 month. I got `Segmentation fault (core dumped)` for version v0.7.2-dev3666 eae707d54c

**Reply by Mytherin** [reply]
Hm, that's not ideal. Would you be open to sharing the data set with me privately so I can reproduce this?

Alternatively, could you perhaps rerun with `SET preserve_insertion_order=false`, and with `read_csv_auto(..., parallel=False)`/`read_csv_auto(..., parallel=True)` so we could narrow down the cause of the issue?

**Reply by albahhar-ahmed** [reply]
I cannot share the data unfortunately. I wanted to write a script to generate a similar data structure and share it with you but things have been busy these few days. I'll try your suggestions and get back to you when things slow down a bit.

Here is something a noticed today. I was suspecting that the pattern matching was taking long and its seems to be the case. To make things even worse, I'm using `**` instead of `*` so it does a full depth recursive search and if the folder structure is deep enough, it might be what's causing the seg fault (I'm just guessing here). Basically, I should do
```
read_csv_auto('root/*/*/*/*.csv')
```
instead of
```
read_csv_auto('root/**/**/**/*.csv')
```

I think the long pause before the progress bar shows up is the time it takes to iterate over the paths. Also something I left out because I thought it was irrelevant to the problem: the folder structure I have is year/month/day/level1/level2 where I have ~500 directories in level1 and ~200 directories in level. Therefore, the tree is deep and large so suspecting issues with the paths seems reasonable to me. I hope this helps and I'll get back to you sometime soon :wave:

**Comment by l1t1** [comment]
try to copy less files in one batch, for example, limit to one day's files

**Comment by l1t1** [comment]
https://github.com/duckdb/duckdb/pull/7490 this merge？

**Comment by l1t1** [comment]
you can mv all files to yyyymmddfiles. csv in a directory, then use cat *.csv to merge them in one file, then to copy it to duckdb

**Reply by albahhar-ahmed** [reply]
Although this will probably work, this is not an ideal solution. I have access to the folder structure for now but that will not always be the case. The data originally lives in a s3 bucket and is updated regularly so I'd rather avoid downloading them and restructuring (and would likely not be possible in the future due to file size, bandwidth, and cost) to yyyymmdd format. It'd would great if I can do something like

```sql
CREATE TABLE tbl AS SELECT * FROM read_csv_auto('s3://root/**/**/**/*.csv');
``` 

but thanks for the help. It's much appreciated :pray: .

**Comment by l1t1** [comment]
https://github.com/benkeen/generatedata can generate csv for test

**Comment by l1t1** [comment]
learn from https://unix.stackexchange.com/questions/94207/recursively-add-a-file-to-all-sub-directories

```
kylin@kylin:~/test$ echo 1 >m.csv
kylin@kylin:~/test$ cat m.csv
1
kylin@kylin:~/test$ mkdir -p {2000..2023}/{01..12}
kylin@kylin:~/test$ find . -type d -exec sh -c 'for d; do cp -p ~/test/m.csv "$d/file"; done' _ {} +
kylin@kylin:~/test$ ls 2000
01  02  03  04  05  06  07  08  09  10  11  12  file
kylin@kylin:~/test$ cat 2000/file
1
```

**Comment by l1t1** [comment]
reading 9554 files is ok
```
kylin@kylin:~/test$ mkdir -p {2000..2023}/{01..12}/{01..31}
kylin@kylin:~/test$ find . -type d -exec sh -c 'for d; do cp -p ~/test/m.csv "$d/2.csv"; done' _ {} +

kylin@kylin:~/test$ ~/ch/duckdb
v0.8.0 e8e4cea5ec
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D SELECT * FROM read_csv_auto('~/test/**/*.csv' ,filename=true) limit 30;
┌─────────┬───────────────────────────────────┐
│ column0 │             filename              │
│  int64  │              varchar              │
├─────────┼───────────────────────────────────┤
│       1 │ /home/kylin/test/m.csv            │
│       1 │ /home/kylin/test/2.csv            │
│       1 │ /home/kylin/test/2007/m.csv       │
│       1 │ /home/kylin/test/2007/2.csv       │
│       1 │ /home/kylin/test/2007/09/m.csv    │
│       1 │ /home/kylin/test/2007/09/2.csv    │
│       1 │ /home/kylin/test/2007/09/21/2.csv │
│       1 │ /home/kylin/test/2007/09/28/2.csv │
│       1 │ /home/kylin/test/2007/09/31/2.csv │
│       1 │ /home/kylin/test/2007/09/26/2.csv │
│       1 │ /home/kylin/test/2007/09/09/2.csv │
│       1 │ /home/kylin/test/2007/09/15/2.csv │
│       1 │ /home/kylin/test/2007/09/03/2.csv │
│       1 │ /home/kylin/test/2007/09/24/2.csv │
│       1 │ /home/kylin/test/2007/09/30/2.csv │
│       1 │ /home/kylin/test/2007/09/11/2.csv │
│       1 │ /home/kylin/test/2007/09/18/2.csv │
│       1 │ /home/kylin/test/2007/09/07/2.csv │
│       1 │ /home/kylin/test/2007/09/08/2.csv │
│       1 │ /home/kylin/test/2007/09/16/2.csv │
│       1 │ /home/kylin/test/2007/09/04/2.csv │
│       1 │ /home/kylin/test/2007/09/14/2.csv │
│       1 │ /home/kylin/test/2007/09/12/2.csv │
│       1 │ /home/kylin/test/2007/09/25/2.csv │
│       1 │ /home/kylin/test/2007/09/20/2.csv │
│       1 │ /home/kylin/test/2007/09/01/2.csv │
│       1 │ /home/kylin/test/2007/09/13/2.csv │
│       1 │ /home/kylin/test/2007/09/29/2.csv │
│       1 │ /home/kylin/test/2007/09/27/2.csv │
│       1 │ /home/kylin/test/2007/09/22/2.csv │
├─────────┴───────────────────────────────────┤
│ 30 rows                           2 columns │
└─────────────────────────────────────────────┘
D .timer on
D SELECT count(*) FROM read_csv_auto('~/test/**/*.csv') ;
┌──────────────┐
│ count_star() │
│    int64     │
├──────────────┤
│         9554 │
└──────────────┘
Run Time (s): real 0.320 user 0.164000 sys 0.332000
D SELECT sum(column0) FROM read_csv_auto('~/test/**/*.csv') ;
┌──────────────┐
│ sum(column0) │
│    int128    │
├──────────────┤
│         9554 │
└──────────────┘
Run Time (s): real 0.320 user 0.176000 sys 0.316000
D 
```

**Comment by l1t1** [comment]
my test shows one `**` like `read_csv_auto('~/test/**/*.csv')` can do search recursive
